common/trinary/flex_trit.c

```diff
@@ -187,3 +198,107 @@ size_t flex_trits_from_trytes(flex_trit_t *to_flex_trits, size_t to_len,
+  for (int i = 0, j = 0; num_trits; i++, j++) {
+    max_trits = MIN(NUM_TRITS_PER_FLEX_TRIT, num_trits);
+    flex_trits_to_trits(shifter.trits + offset, max_trits, &flex_trits[i],
+                        max_trits, max_trits);
+    num_trits -= max_trits;
+    offset += max_trits;
+    if (offset < 5 && num_trits) {
+      i++;
+      max_trits = MIN(NUM_TRITS_PER_FLEX_TRIT, num_trits);
+      flex_trits_to_trits(shifter.trits + offset, max_trits, &flex_trits[i],
+                          max_trits, max_trits);
+      num_trits -= max_trits;
+      offset += max_trits;
+    }
+    trits_for_byte = MIN(4, (offset - 1));
+    bytes[j] = trits_to_byte(shifter.trits, 0, trits_for_byte);
+#if BYTE_ORDER == LITTLE_ENDIAN
+    shifter.val = shifter.val >> 40;
+#elif BYTE_ORDER == BIG_ENDIAN
+    shifter.val = shifter.val << 40;
+#endif
+    offset -= 5;
+  }
+#elif defined(FLEX_TRIT_ENCODING_5_TRITS_PER_BYTE)
+  size_t num_bytes = min_bytes(num_trits);
+  memcpy(bytes, flex_trits, num_bytes);
+  size_t residual = num_trits % 5;
+  if (residual) {
+    trit_t last_byte[5] = {0};
+    size_t index = num_bytes - 1;
+    byte_to_trits(flex_trits[index], last_byte, 4);
+    bytes[index] = trits_to_byte(last_byte, 0, residual - 1);
+  }
+#endif
+  return num_trits;
+}
+
+size_t flex_trits_from_bytes(flex_trit_t *to_flex_trits, size_t to_len,
+                             const byte_t *bytes, size_t len,
+                             size_t num_trits) {
+  // Bounds checking
+  if (num_trits > len || num_trits > to_len) {
+    return 0;
+  }
+  memset(to_flex_trits, FLEX_TRIT_NULL_VALUE, num_flex_trits_for_trits(to_len));
+#if defined(FLEX_TRIT_ENCODING_1_TRIT_PER_BYTE)
+  size_t num_bytes = min_bytes(num_trits);
+  bytes_to_trits(bytes, num_bytes, to_flex_trits, num_trits);
+#elif defined(FLEX_TRIT_ENCODING_3_TRITS_PER_BYTE) || \
+    defined(FLEX_TRIT_ENCODING_4_TRITS_PER_BYTE)
+  union _shifter {
+    uint64_t val;
+    trit_t trits[8];
+  };
+  union _shifter shifter = {0};
+  size_t offset = 0;
+  size_t max_trits, trits_for_byte;
+  for (int i = 0, j = 0; num_trits; j++) {
+    max_trits = MIN(5, num_trits);
+    if (offset < NUM_TRITS_PER_FLEX_TRIT) {
+      byte_to_trits(bytes[i], shifter.trits + offset, max_trits - 1);
+      offset += max_trits;
+      i++;
+    }
+    trits_for_byte = MIN(max_trits, NUM_TRITS_PER_FLEX_TRIT);
+    flex_trits_from_trits(&to_flex_trits[j], num_trits, shifter.trits,
+                          trits_for_byte, trits_for_byte);
+    num_trits -= trits_for_byte;  // MIN(max_trits, NUM_TRITS_PER_FLEX_TRIT);
+    shifter.val = shifter.val >> (trits_for_byte << 3);
+    offset -= trits_for_byte;
+  }
+#elif defined(FLEX_TRIT_ENCODING_5_TRITS_PER_BYTE)
+  size_t num_bytes = min_bytes(num_trits);
+  memcpy(to_flex_trits, bytes, num_bytes);
+  size_t residual = num_trits % 5;
+  if (residual) {
+    trit_t last_byte[5] = {0};
+    size_t index = num_bytes - 1;
+    byte_to_trits(bytes[index], last_byte, 4);
```

Comment:
Can't we just do `bytes[index] = FLEX_TRIT_NULL_VALUE` here?

Reply:
(Same goes for the other places?)

Reply:
Not really, the point is to leave only the last 'residual' trits from the last byte and eliminate the trailing trits. Also, bytes[index] contains n trits, not individual trits, so it makes little sense in this case to assign FLEX_TRIT_NULL_VALUE here.